android/runtime/v8/src/native/modules/EvaluateModule.cpp

```diff
@@ -0,0 +1,351 @@
+		Local<Context> moduleContext = Context::New(isolate);
+		Context::Scope contextScope(moduleContext);
+
+		// Obtain module global context.
+		Local<Object> moduleGlobal = moduleContext->Global();
+
+		// Set security token from previous context to access properties.
+		moduleContext->SetSecurityToken(context->GetSecurityToken());
+
+		if (!contextObj.IsEmpty()) {
+			Local<Array> keys = contextObj->GetOwnPropertyNames(context).ToLocalChecked();
+
+			// Context object has been provided, set context properties in module global context.
+			for (int i = 0; i < keys->Length(); i++) {
+				Local<String> key = keys->Get(i).As<String>();
+				Local<Value> value = contextObj->Get(key);
+
+				if (!key.IsEmpty() && !value.IsEmpty()) {
+					moduleGlobal->Set(moduleContext, key, value);
+				}
+			}
+		}
+
+		// Instantiate module and process imports via `ModuleCallback`.
+		module->InstantiateModule(moduleContext, ModuleCallback);
+
+		if (module->GetStatus() == Module::kErrored) {
+
+			// Throw any exceptions from module instantiation.
+			isolate->ThrowException(module->GetException());
+		}
+		if (tryCatch.HasCaught()) {
+			V8Util::openJSErrorDialog(isolate, tryCatch);
+			V8Util::reportException(isolate, tryCatch, true);
+			return;
+		}
+
+		// Execute module, obtaining a result.
+		MaybeLocal<Value> maybeResult = module->Evaluate(moduleContext);
+
+		if (module->GetStatus() == Module::kErrored) {
+
+			// Throw any exceptions from module evaluation.
+			isolate->ThrowException(module->GetException());
+		}
+		if (tryCatch.HasCaught()) {
+			V8Util::openJSErrorDialog(isolate, tryCatch);
+			V8Util::reportException(isolate, tryCatch, true);
+			return;
+		}
+		if (maybeResult.IsEmpty()) {
+
+			// NOTE: This should never happen.
+			LOGW(TAG, "Evaluating module '%s' returned no result.", *String::Utf8Value(isolate, filename));
+
+			// No result found, return undefined.
+			args.GetReturnValue().Set(Undefined(isolate));
+			return;
+		}
+		Local<Value> result = maybeResult.ToLocalChecked();
+
+		if(result->IsPromise()) {
+			Local<Promise> promise = result.As<Promise>();
+
+			// When top-level-await is enabled, modules return a `Promise`.
+			// Wait for the promise to fulfill before obtaining module exports.
+			if (promise->State() == Promise::kPending) {
+
+				// Allow promise to fulfill.
+				isolate->PerformMicrotaskCheckpoint();
+			}
+			if (module->GetStatus() == Module::kErrored) {
+
+				// Throw any exceptions from promise.
+				isolate->ThrowException(module->GetException());
+			}
+			if (tryCatch.HasCaught()) {
+				V8Util::openJSErrorDialog(isolate, tryCatch);
+				V8Util::reportException(isolate, tryCatch, true);
+				return;
+			}
+
+			// Obtain module exports as result.
+			result = module->GetModuleNamespace().As<Object>();
+		}
+
+		// Return result.
+		args.GetReturnValue().Set(result);
+	}
+
+	void EvaluateModule::RunAsScript(const FunctionCallbackInfo<Value>& args)
+	{
+		Isolate* isolate = args.GetIsolate();
```

Comment:
Most of the body in the script/modules codepaths is the same, curious if you could extract and re-use some functions/code to avoid the duplication. Looks to me that the ScriptOrigin and the actual methods/types called for compiling/executing are the main differences.